sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1349,6 +1353,12 @@ object ConvertToLocalRelation extends Rule[LogicalPlan] {
 
     case Limit(IntegerLiteral(limit), LocalRelation(output, data, isStreaming)) =>
       LocalRelation(output, data.take(limit), isStreaming)
+
+    case Filter(condition, LocalRelation(output, data, isStreaming))
+        if !hasUnevaluableExpr(condition) =>
```

Comment:
If the condition is non-deterministic, the values will be always true after the plans are optimized. The DataFrame with non-deterministic filters will always return the same result.

Reply:
We blocks all these optimization in the other optimization rules, e.g., ConstantFolding. All the non-deterministic expressions are not foldable.

Reply:
ConvertToLocalRelation was already doing this for Project so I assumed it's OK. what I did is exactly how Project(LocalRelation) works.

Reply:
I am fine about introducing this change since we already did it in Project(LocalRelation).

Reply:
I suppose it is fine in this case. The only thing is that it violates the contract of the optimizer: it should not change the results of a query.